File: libs/python/book_creator_schemas/models/messages.py
"""Agent message and critique representations."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID, uuid4

from pydantic import BaseModel, Field

from ..enums import AgentRole, BookStage, CritiqueSeverity


class CritiqueNote(BaseModel):
    """Feedback issued by a critic agent."""

    id: UUID = Field(default_factory=uuid4)
    severity: CritiqueSeverity = CritiqueSeverity.INFO
    summary: str = Field(..., min_length=1, max_length=500)
    details: str = Field(...)
    target_reference: str = Field(
        ..., description="Identifier of the element the critique refers to (chapter, fact, etc.)"
    )
    applied: bool = False


class AgentMessage(BaseModel):
    """Conversation trace for transparency and auditing."""

    id: UUID = Field(default_factory=uuid4)
    project_id: UUID
    stage: BookStage
    role: AgentRole
    content: str
    created_at: datetime = Field(default_factory=datetime.utcnow)
    critiques: list[CritiqueNote] = Field(default_factory=list)
    resulting_artifact_ids: list[UUID] = Field(
        default_factory=list, description="Artifacts produced or updated by this message"
    )

File: libs/ts/src/schemas/messages.ts
import { z } from "zod";
import { AgentRole, BookStage, CritiqueSeverity } from "./enums";

const uuid = z.string().uuid();

export const CritiqueNote = z.object({
  id: uuid,
  severity: CritiqueSeverity.default("info"),
  summary: z.string().min(1).max(500),
  details: z.string().min(1),
  target_reference: z.string().min(1),
  applied: z.boolean().default(false),
});
export type CritiqueNote = z.infer<typeof CritiqueNote>;

export const AgentMessage = z.object({
  id: uuid,
  project_id: uuid,
  stage: BookStage,
  role: AgentRole,
  content: z.string().min(1),
  created_at: z.string().datetime(),
  critiques: z.array(CritiqueNote).default([]),
  resulting_artifact_ids: z.array(uuid).default([]),
});
export type AgentMessage = z.infer<typeof AgentMessage>;
